runtime.md

```diff
@@ -37,23 +37,66 @@ This is provided so that consumers can find the container's configuration and ro
-   The container could also error out or crash.
-8. The container is destroyed by undoing the steps performed during create phase (step 2).
-9. The poststop hooks are invoked by the runtime and errors, if any, are logged.
-10. The state.json file associated with the container is removed and the return code of the container's user specified process is returned or logged.
+1. OCI compliant runtime is invoked with a reference to the location of the bundle.
+   How this reference is passed to the runtime is an implementation detail.
+2. The container's runtime environment MUST be created according to the configuration in `config.json` and `runtime.json`.
+   Any updates to `config.json` or `runtime.json` after container is running MUST not affect the container.
+3. The container's `state.json` file MUST be written to the filesystem.
+4. The prestart hooks MUST be invoked by the runtime.
+   If any prestart hook fails, then the container MUST be stopped and the lifecycle continues at step 8.
+5. The user specified process MUST be executed in the container.
+6. The poststart hooks MUST be invoked by the runtime.
+   If any poststart hook fails, then the container MUST be stopped and the lifecycle continues at step 8.
+7. Additional actions such as pausing the container, resuming the container or signaling the container MAY be performed using the runtime interface.
+   The container MAY also error out, exit or crash.
+8. The container MUST be destroyed by undoing the steps performed during create phase (step 2).
+9. The poststop hooks MUST be invoked by the runtime and errors, if any, MAY be logged.
+10. The `state.json` file associated with the container MUST be removed and the return code of the container's user specified process MUST be returned or logged.
 
 Note: The lifecycle is a WIP and it will evolve as we have more use cases and more information on the viability of a separate create phase.
 
+## Operations
+
+OCI compliant runtimes MUST support the following operations, unless the operation is not supported by the base operating system.
+
+### Errors
+In cases where the specified operation generates an error, this specification does not mandate how, or even if, that error is returned or exposed to the user of an implementation.
+Unless otherwise stated, generating an error MUST leave the state of the environment as if the operation were never attempted - modulo any possible trivial ancillary changes such as logging.
+
+### Start
+
+Using the `config.json` and `runtime.json` files, along with the filesystem path referenced in the `config.json`, this operation MUST create a new container.
+This includes creating the relevant namespaces, cgroups and configuring the appropriate capabilities for the container.
+A new process in the container's PID namespace MUST be created as specified by the `config.json` file.
```

Comment:
On Thu, Jan 07, 2016 at 10:40:43AM -0800, Doug Davis wrote:
> > +A new process in the container's PID namespace MUST be created as specified by the `config.json` file.
>
> Well, since the definition of a "container" varies between people I
> wanted to be clear that as part of starting a container we start a
> new process, and that process isn't just running **anywhere** (like
> on the host) its running in the container's PID namespace.  Might be
> anal to say it, but unless its wrong I think its better to be clear.

I agree that dipping into informative concrete examples is useful, but
I'd rather avoid normative language here when the process and PID
namespace settings are already fully spec'ed in other places (and if
the namespace for config.json's process is not clear, that seems like
something that should be addressed in [1], [2], or both).  How about:

  All of the settings defined in [`config.json`](config.md) and
  [`runtime.json`](runtime.md) are applied in this operation.  For
  example, creating namespaces, setting resource limits, mounting
  filesystems, creating devices, launching a container process,
  setting capabilities for that process, and executing the configured
  command all happen here.

[1]: https://github.com/opencontainers/specs/blob/6a6ba6775567e726c405456e74b0e52350104eda/config.md#process-configuration
[2]: https://github.com/opencontainers/specs/blob/6a6ba6775567e726c405456e74b0e52350104eda/runtime-config-linux.md#namespaces